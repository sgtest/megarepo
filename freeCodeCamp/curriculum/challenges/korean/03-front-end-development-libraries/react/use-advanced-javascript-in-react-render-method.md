---
id: 5a24c314108439a4d4036183
title: 리액트 렌더 메서드에 자바스크립트 사용하기
challengeType: 6
forumTopicId: 301415
dashedName: use-advanced-javascript-in-react-render-method
---

# --description--

이전 과제들에서, props에 접근하거나 props를 전달하고, state에 접근하며, 코드에 주석을 삽입하고, 가장 최근에는 컴포넌트를 스타일링하는 등의 작업을 위해 중괄호 `{ }`를 사용하여 JSX 코드에 자바스크립트 코드를 주입하는 방법을 배웠습니다. 모두 JSX에 자바스크립트를 넣는 일반적인 사용 방법이지만, 리액트 컴포넌트에 자바스크립트 코드를 사용할 수 있는 유일한 방법은 아닙니다.

중괄호를 ***사용하지 않고*** `return`문 앞에 있는, `render` 메서드에 자바스크립트를 넣을 수 있습니다. 왜냐하면 아직 JSX 코드 안에 있지 않기 때문입니다. `return`문 *안에서* 변수를 사용하고 싶을 때는, 중괄호에 변수 이름을 넣어주면 됩니다.

# --instructions--

코드 편집기의 렌더(`render`) 메서드에는 1980년대 장난감인 'Magic Eight Ball'의 대답 20개를 포함한 배열이 있습니다. 버튼 클릭 이벤트는 `ask` 메서드에 연결되어 있어서, 버튼을 클릭할 때마다 무작위 숫자가 생성되고, state의 `randomIndex`에 저장됩니다. 52번째 줄에서 문자열 `change me!`를 삭제하고 `answer` const를 재할당하여, 컴포넌트가 업데이트될 때마다 코드가 `possibleAnswers` 배열의 다른 인덱스에 무작위로 접근하도록 하세요. 마지막으로 `answer` const를 `p` 태그에 넣어주세요.

# --hints--

`MagicEightBall` 컴포넌트는 존재하고, 페이지에 렌더링되어야 합니다.

```js
assert.strictEqual(
  Enzyme.mount(React.createElement(MagicEightBall)).find('MagicEightBall')
    .length,
  1
);
```

`MagicEightBall`의 첫 번째 자식은 `input` 요소여야 합니다.

```js
assert.strictEqual(
  Enzyme.mount(React.createElement(MagicEightBall))
    .children()
    .childAt(0)
    .name(),
  'input'
);
```

`MagicEightBall`의 세 번째 자식은 `button`요소여야 합니다.

```js
assert.strictEqual(
  Enzyme.mount(React.createElement(MagicEightBall))
    .children()
    .childAt(2)
    .name(),
  'button'
);
```

`MagicEightBall`의 state는 `userInput` 속성과 `randomIndex` 속성이 모두 빈 문자열 값으로 설정되어 초기화되어야 합니다.

```js
assert(
  Enzyme.mount(React.createElement(MagicEightBall)).state('randomIndex') ===
    '' &&
    Enzyme.mount(React.createElement(MagicEightBall)).state('userInput') === ''
);
```

`MagicEightBall`이 처음 DOM에 마운트될 때는 빈 `p` 요소를 반환해야 합니다.

```js
assert(
  Enzyme.mount(React.createElement(MagicEightBall)).find('p').length === 1 &&
    Enzyme.mount(React.createElement(MagicEightBall)).find('p').text() === ''
);
```

`input` 요소에 텍스트를 입력하고 버튼을 클릭하면, `MagicEightBall` 컴포넌트는 `possibleAnswers` 배열에서 무작위 요소를 포함한 `p` 요소를 반환해야 합니다.

```js
(() => {
  const comp = Enzyme.mount(React.createElement(MagicEightBall));
  const simulate = () => {
    comp.find('input').simulate('change', { target: { value: 'test?' } });
    comp.find('button').simulate('click');
  };
  const result = () => comp.find('p').text();
  const _1 = () => {
    simulate();
    return result();
  };
  const _2 = () => {
    simulate();
    return result();
  };
  const _3 = () => {
    simulate();
    return result();
  };
  const _4 = () => {
    simulate();
    return result();
  };
  const _5 = () => {
    simulate();
    return result();
  };
  const _6 = () => {
    simulate();
    return result();
  };
  const _7 = () => {
    simulate();
    return result();
  };
  const _8 = () => {
    simulate();
    return result();
  };
  const _9 = () => {
    simulate();
    return result();
  };
  const _10 = () => {
    simulate();
    return result();
  };
  const _1_val = _1();
  const _2_val = _2();
  const _3_val = _3();
  const _4_val = _4();
  const _5_val = _5();
  const _6_val = _6();
  const _7_val = _7();
  const _8_val = _8();
  const _9_val = _9();
  const _10_val = _10();
  const actualAnswers = [
    _1_val,
    _2_val,
    _3_val,
    _4_val,
    _5_val,
    _6_val,
    _7_val,
    _8_val,
    _9_val,
    _10_val
  ];
  const hasIndex = actualAnswers.filter(
    (answer, i) => possibleAnswers.indexOf(answer) !== -1
  );
  const notAllEqual = new Set(actualAnswers);
  assert(notAllEqual.size > 1 && hasIndex.length === 10);
})();
```

# --seed--

## --after-user-code--

```jsx
var possibleAnswers = [
  'It is certain',
  'It is decidedly so',
  'Without a doubt',
  'Yes, definitely',
  'You may rely on it',
  'As I see it, yes',
  'Outlook good',
  'Yes',
  'Signs point to yes',
  'Reply hazy try again',
  'Ask again later',
  'Better not tell you now',
  'Cannot predict now',
  'Concentrate and ask again',
  "Don't count on it",
  'My reply is no',
  'My sources say no',
  'Outlook not so good',
  'Very doubtful',
  'Most likely'
];
ReactDOM.render(<MagicEightBall />, document.getElementById('root'));
```

## --seed-contents--

```jsx
const inputStyle = {
  width: 235,
  margin: 5
};

class MagicEightBall extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userInput: '',
      randomIndex: ''
    };
    this.ask = this.ask.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }
  ask() {
    if (this.state.userInput) {
      this.setState({
        randomIndex: Math.floor(Math.random() * 20),
        userInput: ''
      });
    }
  }
  handleChange(event) {
    this.setState({
      userInput: event.target.value
    });
  }
  render() {
    const possibleAnswers = [
      'It is certain',
      'It is decidedly so',
      'Without a doubt',
      'Yes, definitely',
      'You may rely on it',
      'As I see it, yes',
      'Outlook good',
      'Yes',
      'Signs point to yes',
      'Reply hazy try again',
      'Ask again later',
      'Better not tell you now',
      'Cannot predict now',
      'Concentrate and ask again',
      "Don't count on it",
      'My reply is no',
      'My sources say no',
      'Most likely',
      'Outlook not so good',
      'Very doubtful'
    ];
    const answer = 'change me!'; // Change this line
    return (
      <div>
        <input
          type='text'
          value={this.state.userInput}
          onChange={this.handleChange}
          style={inputStyle}
        />
        <br />
        <button onClick={this.ask}>Ask the Magic Eight Ball!</button>
        <br />
        <h3>Answer:</h3>
        <p>
          {/* Change code below this line */}

          {/* Change code above this line */}
        </p>
      </div>
    );
  }
}
```

# --solutions--

```jsx
const inputStyle = {
  width: 235,
  margin: 5
};

class MagicEightBall extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userInput: '',
      randomIndex: ''
    };
    this.ask = this.ask.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }
  ask() {
    if (this.state.userInput) {
      this.setState({
        randomIndex: Math.floor(Math.random() * 20),
        userInput: ''
      });
    }
  }
  handleChange(event) {
    this.setState({
      userInput: event.target.value
    });
  }
  render() {
    const possibleAnswers = [
      'It is certain',
      'It is decidedly so',
      'Without a doubt',
      'Yes, definitely',
      'You may rely on it',
      'As I see it, yes',
      'Outlook good',
      'Yes',
      'Signs point to yes',
      'Reply hazy try again',
      'Ask again later',
      'Better not tell you now',
      'Cannot predict now',
      'Concentrate and ask again',
      "Don't count on it",
      'My reply is no',
      'My sources say no',
      'Outlook not so good',
      'Very doubtful',
      'Most likely'
    ];
    const answer = possibleAnswers[this.state.randomIndex];
    return (
      <div>
        <input
          type='text'
          value={this.state.userInput}
          onChange={this.handleChange}
          style={inputStyle}
        />
        <br />
        <button onClick={this.ask}>Ask the Magic Eight Ball!</button>
        <br />
        <h3>Answer:</h3>
        <p>{answer}</p>
      </div>
    );
  }
}
```
