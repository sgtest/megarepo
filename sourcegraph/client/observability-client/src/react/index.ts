export * from './TraceSpanProvider'
export * from './useCurrentSpan'
export * from './constants'
