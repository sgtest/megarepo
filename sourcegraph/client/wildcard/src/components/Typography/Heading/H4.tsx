import React from 'react'

import { ForwardReferenceComponent } from '../../../types'

import { Heading, HeadingProps } from './Heading'

type H4Props = HeadingProps

// eslint-disable-next-line id-length
export const H4 = React.forwardRef(({ children, as = 'h4', ...props }, reference) => (
    <Heading as={as} styleAs="h4" {...props} ref={reference}>
        {children}
    </Heading>
)) as ForwardReferenceComponent<'h4', H4Props>
