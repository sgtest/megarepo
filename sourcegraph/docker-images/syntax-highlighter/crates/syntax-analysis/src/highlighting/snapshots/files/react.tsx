const Component: React.FunctionComponent<{}> = () => {
    let name = 'id'
    return (
        <div>
            <h1 id={name}>My Component</h1>
            {[1, 2, 3].map(item => (
                <p key={item}>{item}</p>
            ))}
        </div>
    )
}
