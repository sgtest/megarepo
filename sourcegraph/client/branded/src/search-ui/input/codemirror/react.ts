// Hooks to provide convenience wrappers around CodeMirror extensions

import { type MutableRefObject, type RefObject, useEffect, useRef } from 'react'

import type { EditorView } from '@codemirror/view'

import type { QueryState } from '@sourcegraph/shared/src/search'

import { updateFromQueryState } from './queryState'

/**
 * Update the editor's value, selection and cursor depending on how the search
 * query was changed.
 */
export function useUpdateInputFromQueryState(
    editorRef: RefObject<EditorView>,
    queryState: QueryState,
    startCompletion: (view: EditorView) => void
): void {
    const startCompletionRef = useRef(startCompletion)

    useEffect(() => {
        startCompletionRef.current = startCompletion
    }, [startCompletion])

    useEffect(() => {
        const editor = editorRef.current
        if (!editor) {
            return
        }
        updateFromQueryState(editor, queryState, { startCompletion: startCompletionRef.current })
    }, [editorRef, queryState])
}

/**
 * Helper function creating a ref and updating it when it changes.
 */
export function useMutableValue<T>(value: T): MutableRefObject<T> {
    const valueRef = useRef(value)

    useEffect(() => {
        valueRef.current = value
    }, [value])

    return valueRef
}

/**
 * Helper hook to run the function whenever the provided changes. Unlike
 * useEffect this hook _doesn't_ execute the callback on first render.
 */
export function useOnValueChanged<T = unknown>(value: T, func: () => void): void {
    const previousValue = useRef(value)

    useEffect(() => {
        if (previousValue.current !== value) {
            func()
            previousValue.current = value
        }
    }, [value, func])
}
