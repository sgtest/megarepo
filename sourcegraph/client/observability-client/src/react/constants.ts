import { createContext, type MutableRefObject } from 'react'

import { type Context, ROOT_CONTEXT, trace } from '@opentelemetry/api'

// Store trace context in React ref to avoid re-rendering wrapped components on trace context change.
export type TraceContextRef = MutableRefObject<{ context: Context }>

export const TraceContext = createContext<TraceContextRef>({ current: { context: ROOT_CONTEXT } })

export const REACT_TRACER_NAME = '@sourcegraph/react-manual'
export const reactManualTracer = trace.getTracer(REACT_TRACER_NAME, '0.1')

export enum ReactAttributes {
    ComponentName = 'react.component.name',
    ComponentPropPrefix = 'react.prop',
}
