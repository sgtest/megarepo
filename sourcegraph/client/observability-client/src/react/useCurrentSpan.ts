import { useContext } from 'react'

import { trace, Span } from '@opentelemetry/api'

import { IS_OPEN_TELEMETRY_TRACING_ENABLED, noopSpan } from '../constants'

import { TraceContext } from './constants'

interface UseCurrentSpanResult {
    span?: Span
}

/**
 * Get current OpenTelemetry tracing span from the `TraceSpanProvider` higher in the React tree.
 * Returns `noopSpan` if OpenTelemetry tracing is disabled.
 */
let useCurrentSpan = (): UseCurrentSpanResult => {
    const span = trace.getSpan(useContext(TraceContext).current.context)

    return {
        span,
    }
}

if (!IS_OPEN_TELEMETRY_TRACING_ENABLED) {
    useCurrentSpan = () => ({ span: noopSpan })
}

export { useCurrentSpan }
