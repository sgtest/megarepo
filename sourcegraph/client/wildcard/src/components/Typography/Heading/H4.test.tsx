import { render } from '@testing-library/react'

import { H4 } from './H4'

describe('H4', () => {
    it('should render correctly', () => {
        expect(
            render(
                <H4 alignment="left" mode="single-line">
                    This is H4
                </H4>
            ).asFragment()
        ).toMatchSnapshot()
    })

    it('should render correctly with `as`', () => {
        expect(
            render(
                <H4 alignment="left" mode="single-line" as="p">
                    This is a p
                </H4>
            ).asFragment()
        ).toMatchSnapshot()
    })
})
