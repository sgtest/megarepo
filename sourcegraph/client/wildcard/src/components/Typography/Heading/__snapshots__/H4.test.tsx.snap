// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`H4 > should render correctly 1`] = `
<DocumentFragment>
  <h4
    class="h4 alignLeft singleLine"
  >
    This is H4
  </h4>
</DocumentFragment>
`;

exports[`H4 > should render correctly with \`as\` 1`] = `
<DocumentFragment>
  <p
    class="h4 alignLeft singleLine"
  >
    This is a p
  </p>
</DocumentFragment>
`;
